hydro_lang/src/keyed_stream.rs

```diff
@@ -74,7 +75,50 @@ where
     }
 }
 
-impl<'a, K, V, L: Location<'a>, B, O, R> KeyedStream<K, V, L, B, O, R> {
+/// Helper function to create a KeyedStream from a transformed underlying stream,
+/// updating the metadata to reflect keyed stream type conversions.
+fn keyed_stream_with_updated_metadata<'a, K, U, L, B, O, R>(
+    underlying: Stream<(K, U), L, B, NoOrder, R>,
+) -> KeyedStream<K, U, L, B, O, R>
+where
+    L: Location<'a>,
+    B: Boundedness,
+{
+    {
+        let mut node = underlying.ir_node.borrow_mut();
+        let metadata = node.metadata_mut();
+        metadata.stream_kind = Some(StreamKind::KeyedStream);
+        metadata.is_bounded = B::is_bounded();
+    }
+    KeyedStream::new(underlying)
+}
+
+/// Helper function to create a KeyedStream from a transformed underlying stream,
+/// updating the metadata to reflect keyed stream type conversions where input and output types are the same.
+fn keyed_stream_with_updated_same_type_metadata<'a, K, V, L, B, O, R>(
```

Comment:
This has the exact same signature (modulo renaming) and implementation as above, so I think we can unify?